### PR TITLE
fix: software compositing infinite loop

### PR DESCRIPTION
**Pull request by samuelmaddock**
#### Description of Change

On virtual desktop machines, such as Citrix, GPU compositing may not be available. In certain cases, we're seeing a race condition where the renderer process will get stuck in an infinite loop when calling `LayerTreeView::RequestNewLayerTreeFrameSink`. This aligns with what's reported in https://issues.chromium.org/issues/345275130

1. [LayerTreeView::RequestNewLayerTreeFrameSink](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/widget/compositing/layer_tree_view.cc;l=319;drc=1bae8be84df46ae0825621f386870b2640cbc751) is called.
2. An attempt to [initialize cc::RasterContextProviderWrapper](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/widget/widget_base.cc;l=799-807;drc=a1c86f5b66b3c8430f2ab39e65a519cd31452a92) fails [due to the GPU being disabled.](https://source.chromium.org/chromium/chromium/src/+/main:services/viz/public/cpp/gpu/context_provider_command_buffer.cc;l=133-141;drc=e35c86a08334445bec5b81cdc7a8b8ffa67529f8)
3. `LayerTreeView::DidFailToInitializeLayerTreeFrameSink` is called which [posts a task to try again.](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/widget/compositing/layer_tree_view.cc;l=377-379;drc=1bae8be84df46ae0825621f386870b2640cbc751) The task is performed in an infinite loop without time to process others.
4. [CompositingModeFallbackToSoftware IPC is sent](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/host/compositing_mode_reporter_impl.cc;l=21;drc=e35c86a08334445bec5b81cdc7a8b8ffa67529f8) from the main/browser process, but the infinite task loop starves the RenderThread from ever receiving it.

Credit to @clavin for his thorough work in debugging the flow which leads to the task starvation bug.

Preview of trace where infinite loop occurs. `LayerTreeView::RequestNewLayerTreeFrameSink` is repeatedly called towards the right, leading to an unresponsive renderer.
Screenshot 2024-10-18 at 11 19 40 AM

The patch introduced adds a delay to posting the `LayerTreeView::RequestNewLayerTreeFrameSink` task. This allows the RenderThread time to receive the `CompositingModeFallbackToSoftware` IPC. We'll attempt to upstream this patch soon.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Fix race condition in which WebContents appears as white screen on VDI machines without hardware acceleration capabilities.

### CODE REVIEW

File: patches/chromium/fix_software_compositing_infinite_loop.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+   if (!Platform::Current()->IsGpuRemoteDisconnected()) {
+-    layer_tree_host_->GetTaskRunnerProvider()->MainThreadTaskRunner()->PostTask(
++    // CompositingModeFallbackToSoftware IPC will disable GPU compositing in
++    // RenderThread. Post task with delay to give time to receive this IPC and
++    // prevent infinite loop of retries for software renderers.
++    // https://issues.chromium.org/345275130
++    layer_tree_host_->GetTaskRunnerProvider()->MainThreadTaskRunner()->PostDelayedTask(
+         FROM_HERE, base::BindOnce(&LayerTreeView::RequestNewLayerTreeFrameSink,
+-                                  weak_factory_.GetWeakPtr()));
++                                  weak_factory_.GetWeakPtr()), base::Milliseconds(10));
```
Comment (clavin) [inline patches/chromium/fix_software_compositing_infinite_loop.patch:28]: The [docs on `SequencedTaskRunner`](https://source.chromium.org/chromium/chromium/src/+/main:base/task/sequenced_task_runner.h;l=127;drc=04989307509c78d67b34f2282ca549397719d046) make me think that this delay could potentially be as low as 1 nanosecond and still give the other task priority. Not sure though.

Comment (samuelmaddock) [reply patches/chromium/fix_software_compositing_infinite_loop.patch:28]: I tested with 1ms delay which seems to also fix the issue. Throttling the timer may reduce CPU usage slightly in cases where it fails multiple times in a row.

When set to 10ms, I saw `LayerTreeView::DidFailToInitializeLayerTreeFrameSink` posted 4 times. When set to 1ms, I only saw 2 occurrences. I'll leave it up to you when creating the CL upstream. The Chromium folks may also have thoughts as to what an appropriate delay might be.

**Comment by deepak1556** [comment]
Thanks for the detailed write up, I am good with the change here but curious to check on an alternative path, are there issues in running with `EarlyEstablishGpuChannel` in the sync path, seems like it should catch this block https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/widget/compositing/layer_tree_view.cc;l=371-376 ?

Eventually the async path with early channel establish might become the default based on this thread https://issues.chromium.org/issues/345275130

**Comment by samuelmaddock** [comment]
> Thanks for the detailed write up, I am good with the change here but curious to check on an alternative path, are there issues in running with `EarlyEstablishGpuChannel` in the sync path, seems like it should catch this block https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/widget/compositing/layer_tree_view.cc;l=371-376 ?

When stepping through with a debugger, the conditional always passed and continued posting the task to call `RequestNewLayerTreeFrameSink`. In our case, this bug occurs upon initial page load in the renderer process. I believe the early return for `IsGpuRemoteDisconnected()` catches the edge case while the application is shutting down.

> Eventually the async path with early channel establish might become the default based on this thread https://issues.chromium.org/issues/345275130

This bug seemed to persist even after enabling `EstablishGpuChannelAsync` via blink features. Including this fix continued to work with the flag enabled.

**Comment by deepak1556** [comment]
> In our case, this bug occurs upon initial page load in the renderer process. I believe the early return for IsGpuRemoteDisconnected() catches the edge case while the application is shutting down.

Thanks for checking, I was expecting the early call would push the channel establishing when render thread gets created https://source.chromium.org/chromium/chromium/src/+/main:content/renderer/render_thread_impl.cc;l=545-553 and failure there would set the gpu channel disconnection https://source.chromium.org/chromium/chromium/src/+/main:services/viz/public/cpp/gpu/gpu.cc;l=425-426 well before the `RequestNewLayerTreeFrameSink` call to avoid the loop. If that worked it would have been a nice option.

> This bug seemed to persist even after enabling EstablishGpuChannelAsync via blink features.

Oh seems a bit different from this observation https://issues.chromium.org/issues/345275130#comment7 would be good to report this so that it is captured in the async path.

**Pull request merged**